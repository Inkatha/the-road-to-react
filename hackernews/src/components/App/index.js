import React, { Component } from 'react';
import './index.css';

import {
  Button,
} from '../Buttons';

import Table from '../Table';

import Search from '../Search';

import {
  DEFAULT_QUERY,
  DEFAULT_PAGE,
  DEFAULT_HPP,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP
} from '../../constants';

const updateSearchTopstoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];
  
  const updatedHits = [
    ...oldHits,
    ...hits
  ];

  return {
    results: {
      ...results, 
      [searchKey]: { hits: updatedHits, page }
    },
      isLoading: false,
  }
}

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      isLoading: false,
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    this.setState(updateSearchTopstoriesState(hits, page))

    
  }

 

  fetchSearchTopStories(searchTerm, pageNumber) {
    this.setState({ isLoading: true });

    const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${pageNumber}&${PARAM_HPP}${DEFAULT_HPP}`;
    fetch(url)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result));
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm })
    this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
  }



  onDismiss(id) {

    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];
    
    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm })
    
    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
    }
    
    event.preventDefault();
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  render() {
    const { 
      searchTerm, 
      results,
      searchKey,
      isLoading,
    } = this.state;

    const page = (
      results && 
      results[searchKey] &&
      results[searchKey].page
    ) || 0;

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search 
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>
        <Table
          list={list}
          onDismiss={this.onDismiss}
        />
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}>
              More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }
}


const Loading = () => 
  <div>
    <i className="fa fa-spinner fa-pulse fa-3x fa-fw"></i>
    <span class="sr-only">Loading...</span>
  </div>


const withLoading = (Component) => ({ isLoading, ...rest }) => 
  isLoading ? <Loading /> : <Component { ...rest } />


const ButtonWithLoading = withLoading(Button);



export default App;

export {
  Button,
  Search,
  Table
};
