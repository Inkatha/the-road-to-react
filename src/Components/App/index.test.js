import React from 'react';
import ReactDOM from 'react-dom';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App, { SubmitButton, SearchTextBox, Table } from './index';

describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  test('snapshots', () => {
    const component = Renderer.create(
      <App />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('SubmitButton', () => {
  
  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SubmitButton>Give Me More</SubmitButton>, div);
  });

  test('snapshot', () => {
    const component = Renderer.create(
      <SubmitButton>Give Me More</SubmitButton>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: '1', points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: '1', points: 2, objectID: 'z' },
    ]
  }

  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('snapshots', () => {
    const component = Renderer.create(
      <Table {...props} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in the list', () => {
    const element = shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2);
  });
  
});

