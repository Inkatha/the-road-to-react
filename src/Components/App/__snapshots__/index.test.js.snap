exports[`App snapshots 1`] = `
<div
  className="page">
  <div
    className="interactions">
    <form
      onSubmit={[Function]}>
      <input
        onChange={[Function]}
        type="text"
        value="redux" />
      <button
        type="submit">
        Search
      </button>
    </form>
  </div>
  <div
    className="table" />
  <div
    className="interactions">
    <button
      className=""
      onClick={[Function]}
      type="button">
      More
    </button>
  </div>
</div>
`;

exports[`SubmitButton snapshot 1`] = `
<button
  className=""
  onClick={undefined}
  type="button">
  Give Me More
</button>
`;

exports[`Table snapshots 1`] = `
<div
  className="table">
  <div
    className="table-row">
    <span
      style={
        Object {
          "largeColumn": Object {
            "width": "40%",
          },
        }
      }>
      <a
        href={undefined}>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "midColumn": Object {
            "width": "30%",
          },
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}>
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row">
    <span
      style={
        Object {
          "largeColumn": Object {
            "width": "40%",
          },
        }
      }>
      <a
        href={undefined}>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "midColumn": Object {
            "width": "30%",
          },
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}>
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
